pyrefly/lib/alt/expr.rs

```diff
@@ -1255,7 +1309,9 @@ impl<'a, Ans: LookupAnswer> AnswersSolver<'a, Ans> {
                 })
             }
             Expr::StringLiteral(ExprStringLiteral { value: key, .. }) => {
-                TypeInfo::at_facet(base, &FacetKind::Key(key.to_string()), || {
+                let key_str = key.to_string();
+                self.warn_if_optional_typed_dict_key_access(base, &key_str, slice.range(), errors);
```

Comment:
subscript_infer_for_type already has a case for handling TypedDicts, I'd prefer to avoid adding places where we match on the Type struct, so I think we should just add that warning there.

Right now we don't check whether the field is NotRequired, it would be fairly simple to add a condition and emit the warning there. If the narrowed facet exists, that code path will not be hit at all.

https://github.com/facebook/pyrefly/blob/db0d49be8e8424f1ba8e6893436a961dcf8f8a67/pyrefly/lib/alt/expr.rs#L2196

Then the additional logic in type_has_optional_typed_dict_key + warn_if_optional_typed_dict_key_access are no longer required.